pl_examples/basic_examples/cpu_template.py

```diff
@@ -1,53 +1,55 @@
 """
 Runs a model on the CPU on a single node.
 """
-import os
-from argparse import ArgumentParser
-
-import numpy as np
-import torch
+import hydra
 
 import pytorch_lightning as pl
 from pl_examples.models.lightning_template import LightningTemplateModel
+from omegaconf import DictConfig, OmegaConf
+from pl_examples.basic_examples.conf.trainer_conf import cs
+
 
 pl.seed_everything(234)
 
 
-def main(args):
+@hydra.main(config_path="conf", config_name="config")
+def main(cfg: DictConfig):
     """
     Main training routine specific for this project
     :param args:
     """
     # ------------------------
     # 1 INIT LIGHTNING MODEL
     # ------------------------
-    model = LightningTemplateModel(**vars(args))
-
+    model = LightningTemplateModel(**cfg.models)
     # ------------------------
     # 2 INIT TRAINER
     # ------------------------
-    trainer = pl.Trainer.from_argparse_args(args)
+    logger = None
+    profiler = None
+    early_stopping = None
+    checkpoint = None
+    if "logger" in cfg.pl:
+        logger = hydra.utils.instantiate(cfg.pl.logger)
+    if "profiler" in cfg.pl:
+        profiler = hydra.utils.instantiate(cfg.pl.profiler)
+    if "early_stopping" in cfg.pl:
+        early_stopping = hydra.utils.instantiate(cfg.pl.early_stopping)
+    if "checkpoint" in cfg.pl:
+        checkpoint = hydra.utils.instantiate(cfg.pl.checkpoint)
+    trainer = pl.Trainer(
```

Comment:
We might want to create a top level config here as well.
see the example [here](https://hydra.cc/docs/next/tutorials/structured_config/nesting).

This top level config is application specific but we can come up with one that core for PL and show the user how to extend it to add their own configs.

Reply:
Question on cleaning this up, I have an idea but want to know if there is a better way. Basically, I could define a data class for each of these profiler, early stopping, checkpoint and logger, then merge these using a DictConfig subclass like in the example here: https://hydra.cc/docs/next/tutorials/structured_config/defaults. After that, what becomes the best way to instantiate these sub objects like logger/early stopping/profiling in the trainer config and pass it through?

Reply:
This can either be a common base class that initializes all the common PL members from the config, or alternatively we just leave this to the user.
This is a general software engineering question. I think we can try adding a base class and see what it looks like.